src-self-hosted/codegen.zig

```diff
@@ -348,172 +350,182 @@ const Function = struct {
         }
     }
 
-    fn genSetReg(self: *Function, src: usize, comptime arch: Target.Cpu.Arch, reg: Reg(arch), mcv: MCValue) !void {
+    fn genSetReg(self: *Function, src: usize, comptime arch: Target.Cpu.Arch, reg: Reg(arch), mcv: MCValue) error{ CodegenFail, OutOfMemory }!void {
         switch (arch) {
-            .x86_64 => switch (reg) {
-                .rax => switch (mcv) {
-                    .none, .unreach => unreachable,
-                    .immediate => |x| {
-                        // Setting the eax register zeroes the upper part of rax, so if the number is small
-                        // enough, that is preferable.
-                        // Best case: zero
-                        // 31 c0     xor    eax,eax
-                        if (x == 0) {
-                            return self.code.appendSlice(&[_]u8{ 0x31, 0xc0 });
+            .x86_64 => switch (mcv) {
+                .none, .unreach => unreachable,
+                .immediate => |x| {
+                    if (reg.size() != 64) {
+                        return self.fail(src, "TODO decide whether to implement non-64-bit loads", .{});
+                    }
+                    // 32-bit moves zero-extend to 64-bit, so xoring the 32-bit
+                    // register is the fastest way to zero a register.
+                    if (x == 0) {
+                        // The encoding for `xor r32, r32` is `0x31 /r`.
+                        // Section 3.1.1.1 of the Intel x64 Manual states that "/r indicates that the
+                        // ModR/M byte of the instruction contains a register operand and an r/m operand."
+                        //
+                        // R/M bytes are composed of two bits for the mode, then three bits for the register,
+                        // then three bits for the operand. Since we're zeroing a register, the two three-bit
+                        // values will be identical, and the mode is three (the raw register value).
+                        //
+                        if (reg.isExtended()) {
+                            // If we're accessing e.g. r8d, we need to use a REX prefix before the actual operation. Since
+                            // this is a 32-bit operation, the W flag is set to zero. X is also zero, as we're not using a SIB.
+                            // Both R and B are set, as we're extending, in effect, the register bits *and* the operand.
+                            //
+                            // From section 2.2.1.2 of the manual, REX is encoded as b0100WRXB. In this case, that's
+                            // b01000101, or 0x45.
```

Comment:
would it be useful to use binary literals?

Reply:
Not really; explaining it in the comments makes sense to me, but there's no real point in writing it out in the actual code. Either way, you need to either read the comments or the manual to understand it.